Add per namespace encryption section for Vault. (#416)
portworx/pxdocs: content/key-management/vault/pvc-enc/_index.md

{{% content "key-management/shared/other-providers-pvc-encryption" %}}

__Important: Make sure secret `your_secret_key` exists in Vault__

### Encryption per Namespace

> In this method, all the PVCs from a namespace will be encrypted using the same secret key. In order to achieve this we will do the following:

Reviewer: ```suggestion
Using this method, you can encrypt all the PVCs from a namespace using the same secret key. To achieve this, you must do the following:
```

> * In the storage class we will set the following two fields:

Reviewer: ```suggestion
* Create a storage class, inserting the following two fields:
```

>     * **secret: "true"**: This indicates that the PVCs created from this storage class are encrypted.

Reviewer: ```suggestion
    * **secret: “true”**: This field instructs Portworx to encrypt the PVCs created from this storage class.
```

>     * **secret_key: "your_secret_key"**: This indicates that the PVCs using this storage class will use the key **your_secret_key** from Vault

Reviewer: ```suggestion
    * **secret_key: “<your_secret_key>”**: This field specifies which Vault key the PVCs created from this storage class use. 
```

> * We will define **ResourceQuotas** on the namespace in which PVCs are created so that only the above Storage Class will be used.

Reviewer: ```suggestion
* Create a **ResourceQuota** object on the namespace in which PVCs are created restricting any other Storage Class from being used.
```

#### Step 1: Create a Storage Class

> Create a storage class with the `secure` parameter set to `true`.

Reviewer: ```suggestion
Begin configuring namespace-level encryption by creating a storage class with the `secure` parameter set to `true`.
```

```text
kind: StorageClass
apiVersion: storage.k8s.io/v1
metadata:
  name: px-secure-sc
provisioner: kubernetes.io/portworx-volume
parameters:
  secure: "true"
  secret_key: "your_secret_key"
  repl: "3"
```

> __Important: Make sure secret `your_secret_key` exists in Vault__

Reviewer: ```suggestion
{{<info>}}
**Important:** Make sure secret `your_secret_key` exists in Vault
{{</info>}}
```

> To create a **shared encrypted volume** set the `shared` parameter to `true` as well.

Reviewer: ```suggestion
To create a **shared encrypted volume**, set the `shared` parameter to `true` as well.
```

#### Step 2: Create a ResourceQuota object for the PVC's Namespace

> Create a resource quota that disallows any other storage classes to be used in this namespace

Reviewer: ```suggestion
Once you've created your StorageClass, create a resource quota that disallows any other storage classes to be used in this namespace.
```

```text
apiVersion: v1
kind: ResourceQuota
metadata:
  name: dev-storageclass-quotas
  namespace: dev
spec:
  hard:
    <storageclass-1-name>.storageclass.storage.k8s.io/persistentvolumeclaims: 0
    <storageclass-2-name>.storageclass.storage.k8s.io/persistentvolumeclaims: 0
    <storageclass-3-name>.storageclass.storage.k8s.io/persistentvolumeclaims: 0
```

> With the above ResourceQuota definition we are disallowing PVC creations using

Reviewer: ```suggestion
The ResourceQuota definition above forbids PVC creations using
```

storage classes `<storageclass-1-name>, <storageclass-2-name>, <storageclass-3-name>` in the `dev` namespace.

> __Note: In the above ResourceQuota we have excluded the `px-secure-sc` storage class as we want to use it to create encrypted PVCs.__

Reviewer: ```suggestion
{{<info>}}
**Note:** In the above ResourceQuota we have excluded the `px-secure-sc` storage class as we want to use it to create encrypted PVCs.
{{</info>}}
```

#### Step 3: Create an encrypted PVC

> 

Reviewer: ```suggestion
Once you've created your StorageClass and ResourceQuota objects, you're ready to create your encrypted PVCs. Create an encrypted PVC by specifying the encryption-configured namespace and StorageClass.
```
